src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs

```diff
@@ -633,6 +592,75 @@ private static ITaskSeriesTimer StartParameterLogTimer(IRecurrentCommand updateC
+            //   3. A timeout fires.
+            //      a. If throwOnTimeout, we throw the FunctionTimeoutException.
+            //      b. If !throwOnTimeout, wait for the task to complete.
+
+            // Start the invokeTask.
+            Task invokeTask = invoker.InvokeAsync(invokeParameters);
+
+            // A task that will throw a TaskCanceledException if #2 occurs.
+            Task shutdownTask = Task.Delay(-1, functionCancellationTokenSource.Token);
+
+            // A task that will throw when either #1 or #2 occurs.
+            Task<Task> invokeOrShutdownTask = Task.WhenAny(invokeTask, shutdownTask);
+
+            // Combine #1 and #2 with a timeout task (handled by this method).
+            bool isTimeout = await TryHandleTimeoutAsync(invokeOrShutdownTask, throwOnTimeout, timeoutTokenSource.Token,
+                timerInterval, instance, () => functionCancellationTokenSource.Cancel());
+
+            // #2 occurred. If we're going to throwOnTimeout, watch for a timeout while we wait for invokeTask to complete.
+            if (throwOnTimeout && !isTimeout && invokeOrShutdownTask.Result == shutdownTask)
+            {
+                await TryHandleTimeoutAsync(invokeTask, throwOnTimeout, timeoutTokenSource.Token, timerInterval, instance, null);
+            }
+
+            await invokeTask;
+        }
+
+        /// <summary>
+        /// Executes a timeout pattern. Throws an exception if the timeoutToken is canceled before taskToTimeout completes and throwOnTimeout is true.
+        /// </summary>
+        /// <param name="taskToTimeout">The task to run.</param>
+        /// <param name="throwOnTimeout">True if the method should throw an OperationCanceledException if it times out.</param>
+        /// <param name="timeoutToken">The token to watch. If it is canceled, taskToTimeout has timed out.</param>
+        /// <param name="onTimeout">A callback to be executed if a timeout occurs.</param>
+        /// <param name="timeoutInterval">The timeout period. Used only in the exception message.</param>
+        /// <param name="instance">The function instance. Used only in the exceptionMessage</param>
+        /// <returns>True if a timeout occurred. Otherwise, false.</returns>
+        private static async Task<bool> TryHandleTimeoutAsync(Task taskToTimeout, bool throwOnTimeout, CancellationToken timeoutToken,
+            TimeSpan timeoutInterval, IFunctionInstance instance, Action onTimeout)
+        {
+            Task timeoutTask = Task.Delay(-1, timeoutToken);
+            Task completedTask = await Task.WhenAny(taskToTimeout, timeoutTask);
+
+            if (completedTask == timeoutTask)
+            {
+                if (onTimeout != null)
+                {
+                    onTimeout();
+                }
+
+                if (throwOnTimeout)
+                {
+                    // If we need to throw, await the task to throw now. This will bubble up and eventually
```

Comment:
I changed the behavior (I used to await here, then catch a TaskCancledException upstream and throw the FunctionTimeoutException there) -- but I never updated the comment.

Fixed.

---
In reply to: [77361102](https://github.com/Azure/azure-webjobs-sdk/pull/813#discussion_r77361102) [](ancestors = 77361102)